pkg/modinstaller/mod_installer.go

```diff
@@ -216,18 +218,180 @@ func (i *ModInstaller) InstallWorkspaceDependencies(ctx context.Context) (err er
+func (i *ModInstaller) calcChangesForUninstall(oldMod *modconfig.Mod, newMod *modconfig.Mod) ChangeSet {
+	changeset := ChangeSet{}
+	// for every require.mod in current
+	//		if not in updated
+	//			remove
+	//		end if
+	// end for
+	for _, requiredMod := range oldMod.Require.Mods {
+		// check if this mod is still a dependency
+		var modInNew *modconfig.ModVersionConstraint
+		if newMod.Require == nil {
+			modInNew = nil
+		} else {
+			modInNew = newMod.Require.GetModDependency(requiredMod.Name)
+		}
+		if modInNew == nil {
+			changeset = append(changeset, &Change{
+				Operation:   DELETE,
+				OffsetStart: requiredMod.DefRange.Start.Byte,
+				OffsetEnd:   requiredMod.BodyRange.End.Byte,
+				Content:     []byte{},
+			})
+		}
+	}
+	return changeset
+}
+
+// calculates changes required in mod.sp to reflect new installs
+func (i *ModInstaller) calcChangesForInstall(oldMod *modconfig.Mod, newMod *modconfig.Mod) ChangeSet {
+	// for every require.mod in new
+	// 		if not in current
+	//			add - install
+	//		end if
+	// end for
+	// add the new ones
+	if newMod.Require == nil {
+		// return an empty changeset
+		return ChangeSet{}
+	}
+	f := hclwrite.NewEmptyFile()
+	rootBody := f.Body()
+	for _, requiredMod := range newMod.Require.Mods {
+		modInContents := oldMod.Require.GetModDependency(requiredMod.Name)
+		if modInContents == nil {
+			modRequireBlock := rootBody.AppendNewBlock("mod", []string{requiredMod.Name})
+			modRequireBlock.Body().SetAttributeValue("version", cty.StringVal(requiredMod.VersionString))
+		}
+	}
+
+	if len(f.Bytes()) == 0 {
+		return ChangeSet{}
+	}
+
+	return ChangeSet{
+		&Change{
+			Operation:   INSERT,
+			OffsetStart: oldMod.Require.BodyRange.End.Byte - 1,
+			Content:     f.Bytes(),
+		},
+	}
+}
+
+// calculates the changes required in mod.sp to reflect updates
+func (i *ModInstaller) calcChangesForUpdate(oldMod *modconfig.Mod, newMod *modconfig.Mod) ChangeSet {
+	changes := ChangeSet{}
+	// for every require.mod in parsed
+	//		if in workspaceMod
+	//			if parsed.version != workspaceMod.version
+	//				update version field - upgrade
+	//			end if
+	//		end if
+	// end for
+	if newMod.Require == nil {
+		// new mod does not have any require
+		// return an empty change set
+		return ChangeSet{}
+	}
+	for _, requiredMod := range oldMod.Require.Mods {
+		modInUpdated := newMod.Require.GetModDependency(requiredMod.Name)
+		if modInUpdated == nil {
+			continue
+		}
+		if modInUpdated.VersionString != requiredMod.VersionString {
+			changes = append(changes, &Change{
+				Operation:   REPLACE,
+				OffsetStart: requiredMod.VersionRange.Start.Byte,
+				OffsetEnd:   requiredMod.VersionRange.End.Byte,
+				Content:     []byte(fmt.Sprintf("version = \"%s\"", modInUpdated.VersionString)),
+			})
+		}
+	}
+	return changes
+}
+
+// loadModForUpdate parses
+func (i *ModInstaller) loadModForUpdate() (*modconfig.Mod, *ByteSequence, error) {
+	mod, err := parse.LoadModfile(i.workspaceMod.ModPath)
```

Comment:
instead of reloading the modfile (which we have already loaded), can we just clone the pre-update require block from the `workspaceMod` and use that

then this function can be called `loadModAsBytes`